Bug: Query-Scheduler - dequeueSelectNode : panic: runtime error: index out of range [-1]

### What is the bug?

We're investigating a problem detected on one of our Mimir clusters.
We observed a lot of errors from the queriers:

```
rpc error: code = Unknown desc = querier has informed the scheduler it is shutting down
```

Following this, we found that the query schedulers were frequently panicking when managing dequeueing:

```golang
panic: runtime error: index out of range [-1]
goroutine 31 [running]:
github.com/grafana/mimir/pkg/scheduler/queue/tree.(*QuerierWorkerQueuePriorityAlgo).dequeueSelectNode(0x47073d?, 0x7f77ee5f65b8?)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/tree/tree_queue_algo_querier_worker_queue_priority.go:117 +0x5d
github.com/grafana/mimir/pkg/scheduler/queue/tree.(*Node).dequeue(0xc001d206c0)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/tree/multi_algorithm_tree_queue.go:253 +0x107
github.com/grafana/mimir/pkg/scheduler/queue/tree.(*MultiAlgorithmTreeQueue).Dequeue(0xc0008ff0a0?, 0xc002b15220?)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/tree/multi_algorithm_tree_queue.go:94 +0x33
github.com/grafana/mimir/pkg/scheduler/queue.(*queueBroker).dequeueRequestForQuerier(0xc001cdaa50, 0xc0026198c0)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/queue_broker.go:137 +0xc3
github.com/grafana/mimir/pkg/scheduler/queue.(*RequestQueue).trySendNextRequestForQuerier(0xc000a38bb0, 0xc0026198c0)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/queue.go:391 +0x3c
github.com/grafana/mimir/pkg/scheduler/queue.(*RequestQueue).dispatcherLoop(0xc000a38bb0)
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/queue.go:323 +0x62d
created by github.com/grafana/mimir/pkg/scheduler/queue.(*RequestQueue).starting in goroutine 29
	/go/src/github.com/grafana/mimir/pkg/scheduler/queue/queue.go:258 +0x4f
```


Related source code (git tag mimir2.15.2):
https://github.com/grafana/mimir/blob/mimir-2.15.2/pkg/scheduler/queue/tree/tree_queue_algo_querier_worker_queue_priority.go#L117

```golang
    currentNodeName := qa.nodeOrder[qa.currentNodeOrderIndex]
```

I'm trying to understand what's happening here and why we try to dequeue a out of range index, but I'm not sure exactly where to look. 

We cannot say for sure when it started due to log retention.

### How to reproduce it?

It happens frequently, we do not have a way to manually reproduce it.

### What did you think would happen?

No panic and restarting of query scheduler

### What was your environment?

Mimir 2.15.2 running on multiple instances 

### Any additional context to share?

Here is the configuration for both the scheduler and the querier.

Query scheduler config:

```yaml
query_scheduler:
    max_outstanding_requests_per_tenant: 100
    querier_forget_delay: 0s
    grpc_client_config:
        max_recv_msg_size: 209715200
        max_send_msg_size: 419430400
        grpc_compression: snappy
        rate_limit: 0
        rate_limit_burst: 0
        backoff_on_ratelimits: false
        backoff_config:
            min_period: 100ms
            max_period: 10s
            max_retries: 10
        initial_stream_window_size: 63KiB1023B
        initial_connection_window_size: 63KiB1023B
        tls_enabled: true
        tls_cert_path: /etc/ssl/ecdsa-ssl-cert-snakeoil.pem
        tls_key_path: /etc/ssl/ecdsa-ssl-key-snakeoil.pem
        tls_ca_path: /etc/ssl/ecdsa-ssl-cert-snakeoil.pem
        tls_server_name: ""
        tls_insecure_skip_verify: true
        tls_cipher_suites: ""
        tls_min_version: VersionTLS13
        connect_timeout: 5s
        connect_backoff_base_delay: 1s
        connect_backoff_max_delay: 5s
    service_discovery_mode: dns
    ring:
        kvstore:
            store: memberlist
            prefix: mimir/query_scheduler/
            consul:
                host: localhost:8500
                acl_token: '********'
                http_client_timeout: 20s
                consistent_reads: false
                watch_rate_limit: 1
                watch_burst_size: 1
                cas_retry_delay: 1s
            etcd:
                endpoints: []
                dial_timeout: 10s
                max_retries: 10
                tls_enabled: false
                tls_cert_path: ""
                tls_key_path: ""
                tls_ca_path: ""
                tls_server_name: ""
                tls_insecure_skip_verify: false
                tls_cipher_suites: ""
                tls_min_version: ""
                username: ""
                password: ""
            multi:
                primary: ""
                secondary: ""
                mirror_enabled: false
                mirror_timeout: 2s
        heartbeat_period: 15s
        heartbeat_timeout: 1m0s
        instance_id: *******
        instance_interface_names:
            - eth1
        instance_port: 0
        instance_addr: ""
        instance_enable_ipv6: false
    max_used_instances: 0
```
Querier config:
```yaml
querier:
    query_store_after: 23h45m0s
    store_gateway_client:
        tls_enabled: true
        tls_cert_path: /etc/ssl/ecdsa-ssl-cert-snakeoil.pem
        tls_key_path: /etc/ssl/ecdsa-ssl-key-snakeoil.pem
        tls_ca_path: /etc/ssl/ecdsa-ssl-cert-snakeoil.pem
        tls_server_name: ""
        tls_insecure_skip_verify: true
        tls_cipher_suites: ""
        tls_min_version: VersionTLS13
    shuffle_sharding_ingesters_enabled: true
    prefer_availability_zone: ""
    streaming_chunks_per_ingester_series_buffer_size: 256
    streaming_chunks_per_store_gateway_series_buffer_size: 256
    minimize_ingester_requests: true
    minimize_ingester_requests_hedging_delay: 3s
    query_engine: mimir
    enable_query_engine_fallback: true
    max_concurrent: 8
    timeout: 5m0s
    max_samples: 50000000
    default_evaluation_interval: 1m0s
    lookback_delta: 5m0s
    promql_experimental_functions_enabled: false
```
I can provide more logs if needed, as the problem happens frequently.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -66,6 +66,7 @@
 * [BUGFIX] Query-frontend: Fix a panic in monolithic mode caused by a clash in labels of the `cortex_client_invalid_cluster_validation_label_requests_total` metric definition. #11455
 * [BUGFIX] Compactor: Fix issue where `MimirBucketIndexNotUpdated` can fire even though the index has been updated within the alert threshold. #11303
 * [BUGFIX] Distributor: fix old entries in the HA Tracker with zero valued "elected at" timestamp. #11462
+* [BUGFIX] Query-scheduler: Fix issue where deregistered querier goroutines can cause a panic if their backlogged dequeue requests are serviced #11510
 
 ### Mixin
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -66,7 +66,7 @@
 * [BUGFIX] Query-frontend: Fix a panic in monolithic mode caused by a clash in labels of the `cortex_client_invalid_cluster_validation_label_requests_total` metric definition. #11455
 * [BUGFIX] Compactor: Fix issue where `MimirBucketIndexNotUpdated` can fire even though the index has been updated within the alert threshold. #11303
 * [BUGFIX] Distributor: fix old entries in the HA Tracker with zero valued "elected at" timestamp. #11462
-* [BUGFIX] Query-scheduler: Fix issue where deregistered querier goroutines can cause a panic if their backlogged dequeue requests are serviced #11510
+* [BUGFIX] Query-scheduler: Fix issue where deregistered querier goroutines can cause a panic if their backlogged dequeue requests are serviced. #11510
 
 ### Mixin
 

--- a/pkg/scheduler/queue/queue.go
+++ b/pkg/scheduler/queue/queue.go
@@ -27,10 +27,11 @@ const (
 )
 
 var (
-	ErrInvalidTenantID     = errors.New("invalid tenant id")
-	ErrTooManyRequests     = errors.New("too many outstanding requests")
-	ErrStopped             = errors.New("queue is stopped")
-	ErrQuerierShuttingDown = errors.New("querier has informed the scheduler it is shutting down")
+	ErrInvalidTenantID           = errors.New("invalid tenant id")
+	ErrTooManyRequests           = errors.New("too many outstanding requests")
+	ErrStopped                   = errors.New("queue is stopped")
+	ErrQuerierShuttingDown       = errors.New("querier has informed the scheduler it is shutting down")
+	ErrQuerierWorkerDisconnected = errors.New("querier worker has disconnected")
 )
 
 type RequestKey struct {

--- a/pkg/scheduler/queue/queue.go
+++ b/pkg/scheduler/queue/queue.go
@@ -338,7 +338,9 @@ func (q *RequestQueue) dispatcherLoop() {
 			for currentElement != nil {
 				waitingDequeueReq := currentElement.Value.(*QuerierWorkerDequeueRequest)
 				waitingDequeueReq.sendError(ErrStopped)
-				currentElement = currentElement.Next()
+				nextElement := currentElement.Next()
+				q.waitingDequeueRequestsToDispatch.Remove(currentElement)
+				currentElement = nextElement
 			}
 
 			if !q.queueBroker.isEmpty() {

--- a/pkg/scheduler/queue/queue.go
+++ b/pkg/scheduler/queue/queue.go
@@ -117,12 +117,13 @@ type RequestQueue struct {
 	stopRequested chan struct{} // Written to by stop() to wake up dispatcherLoop() in response to a stop request.
 	stopCompleted chan struct{} // Closed by dispatcherLoop() after a stop is requested and the dispatcher has stopped.
 
-	requestsToEnqueue                chan requestToEnqueue
-	requestsSent                     chan *SchedulerRequest
-	requestsCompleted                chan *SchedulerRequest
-	querierWorkerOperations          chan *querierWorkerOperation
-	waitingDequeueRequests           chan *QuerierWorkerDequeueRequest
-	waitingDequeueRequestsToDispatch *list.List
+	requestsToEnqueue                     chan requestToEnqueue
+	requestsSent                          chan *SchedulerRequest
+	requestsCompleted                     chan *SchedulerRequest
+	querierWorkerOperations               chan *querierWorkerOperation
+	waitingDequeueRequests                chan *QuerierWorkerDequeueRequest
+	waitingDequeueRequestsToDispatch      *list.List
+	waitingDequeueRequestsToDispatchCount *atomic.Int64
 
 	// QueryComponentUtilization encapsulates tracking requests from the time they are forwarded to a querier
 	// to the time are completed by the querier or failed due to cancel, timeout, or disconnect.
@@ -238,12 +239,13 @@ func NewRequestQueue(
 		stopRequested: make(chan struct{}),
 		stopCompleted: make(chan struct{}),
 
-		requestsToEnqueue:                make(chan requestToEnqueue),
-		requestsSent:                     make(chan *SchedulerRequest),
-		requestsCompleted:                make(chan *SchedulerRequest),
-		querierWorkerOperations:          make(chan *querierWorkerOperation),
-		waitingDequeueRequests:           make(chan *QuerierWorkerDequeueRequest),
-		waitingDequeueRequestsToDispatch: list.New(),
+		requestsToEnqueue:                     make(chan requestToEnqueue),
+		requestsSent:                          make(chan *SchedulerRequest),
+		requestsCompleted:                     make(chan *SchedulerRequest),
+		querierWorkerOperations:               make(chan *querierWorkerOperation),
+		waitingDequeueRequests:                make(chan *QuerierWorkerDequeueRequest),
+		waitingDequeueRequestsToDispatch:      list.New(),
+		waitingDequeueRequestsToDispatchCount: atomic.NewInt64(0),
 
 		QueryComponentUtilization: queryComponentCapacity,
 		queueBroker:               newQueueBroker(maxOutstandingPerTenant, forgetDelay),
@@ -310,6 +312,7 @@ func (q *RequestQueue) dispatcherLoop() {
 			requestSent := q.trySendNextRequestForQuerier(waitingDequeueReq)
 			if !requestSent {
 				// No requests available for this querier; add it to the list to try later.
+				q.waitingDequeueRequestsToDispatchCount.Inc()
 				q.waitingDequeueRequestsToDispatch.PushBack(waitingDequeueReq)
 			}
 		}
@@ -322,6 +325,7 @@ func (q *RequestQueue) dispatcherLoop() {
 				nextElement := currentElement.Next() // We have to capture the next element before calling Remove(), as Remove() clears it.
 
 				if q.trySendNextRequestForQuerier(call) {
+					q.waitingDequeueRequestsToDispatchCount.Dec()
 					q.waitingDequeueRequestsToDispatch.Remove(currentElement)
 				}
 
@@ -339,6 +343,7 @@ func (q *RequestQueue) dispatcherLoop() {
 				waitingDequeueReq := currentElement.Value.(*QuerierWorkerDequeueRequest)
 				waitingDequeueReq.sendError(ErrStopped)
 				nextElement := currentElement.Next()
+				q.waitingDequeueRequestsToDispatchCount.Dec()
 				q.waitingDequeueRequestsToDispatch.Remove(currentElement)
 				currentElement = nextElement
 			}

--- a/pkg/scheduler/queue/queue_broker.go
+++ b/pkg/scheduler/queue/queue_broker.go
@@ -132,6 +132,11 @@ func (qb *queueBroker) dequeueRequestForQuerier(
 		return nil, nil, qb.tenantQuerierAssignments.queuingAlgorithm.TenantOrderIndex(), ErrQuerierShuttingDown
 	}
 
+	// check if the specific querier-worker connection was deregistered since its dequeue request was submitted
+	if dequeueReq.WorkerID == unregisteredWorkerID {
+		return nil, nil, qb.tenantQuerierAssignments.queuingAlgorithm.TenantOrderIndex(), ErrQuerierWorkerDisconnected
+	}
+
 	var queuePath tree.QueuePath
 	var queueElement any
 	queuePath, queueElement = qb.tree.Dequeue(

--- a/pkg/scheduler/queue/queue_test.go
+++ b/pkg/scheduler/queue/queue_test.go
@@ -329,6 +329,124 @@ func queueConsume(
 	return lastTenantIdx, err
 }
 
+func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.T) {
+	const forgetDelay = 2 * time.Second
+	const testTimeout = 10 * time.Second
+
+	queue, err := NewRequestQueue(
+		log.NewNopLogger(),
+		2,
+		forgetDelay,
+		promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"}),
+		promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"}),
+		promauto.With(nil).NewHistogram(prometheus.HistogramOpts{}),
+		promauto.With(nil).NewSummaryVec(prometheus.SummaryOpts{}, []string{"query_component"}),
+	)
+	require.NoError(t, err)
+
+	// Start the queue service.
+	ctx := context.Background()
+	require.NoError(t, services.StartAndAwaitRunning(ctx, queue))
+
+	// Two queriers connect, allowing us to enqueue requests sharded to only one of the two.
+	querier1Ctx, querier1CtxCancel := context.WithCancelCause(context.Background())
+	querier1Conn := NewUnregisteredQuerierWorkerConn(querier1Ctx, "querier-1")
+	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier1Conn))
+	querier2Conn := NewUnregisteredQuerierWorkerConn(context.Background(), "querier-2")
+	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier2Conn))
+
+	t.Cleanup(func() {
+		// if the test has failed and the queue does not get cleared,
+		// we must send a shutdown signal for the remaining connected querier
+		// or else StopAndAwaitTerminated will never complete.
+		queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
+	})
+
+	// Enqueue a request from a user which would NOT be assigned to querier-1.
+	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
+	// when there are only one or two queriers in the sorted list of connected queriers
+	reqNotSharderToQuerier1 := &SchedulerRequest{
+		Ctx:                       context.Background(),
+		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
+		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
+	}
+	// This will sit in the queue as querier-2 is not await a request yet,
+	// but we need a request in the queue tree to reach the code path
+	// in QuerierWorkerQueuePriorityAlgo.setup where the panic condition is created
+	// by setting the currentNodeOrderIndex to -1.
+	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotSharderToQuerier1, 1, nil))
+
+	// Querier-1 submits a request to dequeue;
+	// it will not receive anything as the only request in the queue is not sharded to querier-1.
+	querier1SubmitDequeueReqWG := sync.WaitGroup{}
+	querier1SubmitDequeueReqWG.Add(1)
+	querier1DequeueReq := NewQuerierWorkerDequeueRequest(querier1Conn, FirstTenant())
+	go func() {
+		defer querier1SubmitDequeueReqWG.Done()
+		_, _, err = queue.AwaitRequestForQuerier(querier1DequeueReq)
+		// This receives one of two errors, it's a race between a context cancel and ErrQuerierWorkerDisconnected
+		require.Error(t, err)
+	}()
+
+	querier1CrashWG := sync.WaitGroup{}
+	querier1CrashWG.Add(1)
+	go func() {
+		defer querier1CrashWG.Done()
+		for queue.waitingDequeueRequestsToDispatch.Len() < 1 {
+			time.Sleep(time.Millisecond * 1)
+		}
+		// Simulate a crash of querier-1 (no graceful shutdown notification).
+		querier1CtxCancel(errors.New("crash"))
+		queue.SubmitUnregisterQuerierWorkerConn(querier1Conn) // normally done in the defers of the grpc loop
+	}()
+
+	// Enqueue another request from a user which would NOT be assigned to querier-1.
+	reqNotSharderToQuerier1 = &SchedulerRequest{
+		Ctx:                       context.Background(),
+		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
+		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
+	}
+	// The enqueuing of any new request triggers an attempt to dispatch requests
+	// to the items in waitingDequeueRequestsToDispatch list and their associated worker conns.
+	//
+	// The dequeue request at the front of that list will be associated with the disconnected worker conn,
+	// which now has a workerID set to unregisteredWorkerID = -1.
+	// Without any check in place to bail out before dequeuing for a workerID with -1,
+	// the QuerierWorkerQueuePriorityAlgo.dequeueSelectNode would panic with index out of range [-1].
+	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotSharderToQuerier1, 1, nil))
+
+	// The dequeue request should be dropped from the waiting list due to the error.
+	assert.Eventually(
+		t,
+		func() bool { return queue.waitingDequeueRequestsToDispatch.Len() == 0 },
+		testTimeout,
+		1*time.Second,
+	)
+
+	// Allow Querier-2 to clear the queue after the disconnected querier-1 is forgotten and a reshard occurs.
+	querier2SubmitDequeueReqWG := sync.WaitGroup{}
+	querier2SubmitDequeueReqWG.Add(1)
+	go func() {
+		defer querier2SubmitDequeueReqWG.Done()
+
+		req, nextTenantIdx, err := queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier2Conn, FirstTenant()))
+		require.NoError(t, err)
+		require.NotNil(t, req)
+
+		req, _, err = queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier2Conn, nextTenantIdx))
+		require.NoError(t, err)
+		require.NotNil(t, req)
+	}()
+
+	querier2SubmitDequeueReqWG.Wait()
+	assert.True(t, queue.queueBroker.isEmpty())
+
+	// These should be long gone but ensure we have no hanging goroutines
+	querier1SubmitDequeueReqWG.Wait()
+	querier1CrashWG.Wait()
+}
+
 func TestRequestQueue_RegisterAndUnregisterQuerierWorkerConnections(t *testing.T) {
 	const forgetDelay = 3 * time.Second
 

--- a/pkg/scheduler/queue/queue_test.go
+++ b/pkg/scheduler/queue/queue_test.go
@@ -366,7 +366,7 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	// Enqueue a request from a user which would NOT be assigned to querier-1.
 	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
 	// when there are only one or two queriers in the sorted list of connected queriers
-	reqNotSharderToQuerier1 := &SchedulerRequest{
+	reqNotShardedToQuerier1 := &SchedulerRequest{
 		Ctx:                       context.Background(),
 		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
 		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
@@ -375,7 +375,7 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	// but we need a request in the queue tree to reach the code path
 	// in QuerierWorkerQueuePriorityAlgo.setup where the panic condition is created
 	// by setting the currentNodeOrderIndex to -1.
-	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotSharderToQuerier1, 1, nil))
+	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
 
 	// Querier-1 submits a request to dequeue;
 	// it will not receive anything as the only request in the queue is not sharded to querier-1.
@@ -402,7 +402,7 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	}()
 
 	// Enqueue another request from a user which would NOT be assigned to querier-1.
-	reqNotSharderToQuerier1 = &SchedulerRequest{
+	reqNotShardedToQuerier1 = &SchedulerRequest{
 		Ctx:                       context.Background(),
 		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
 		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
@@ -412,9 +412,8 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	//
 	// The dequeue request at the front of that list will be associated with the disconnected worker conn,
 	// which now has a workerID set to unregisteredWorkerID = -1.
-	// Without any check in place to bail out before dequeuing for a workerID with -1,
-	// the QuerierWorkerQueuePriorityAlgo.dequeueSelectNode would panic with index out of range [-1].
-	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotSharderToQuerier1, 1, nil))
+	// Attempting to dequeue a request from the tree for an unregisteredWorkerID should fail, but not panic.
+	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
 
 	// The dequeue request should be dropped from the waiting list due to the error.
 	assert.Eventually(

--- a/pkg/scheduler/queue/queue_test.go
+++ b/pkg/scheduler/queue/queue_test.go
@@ -355,13 +355,26 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	querier2Conn := NewUnregisteredQuerierWorkerConn(context.Background(), "querier-2")
 	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier2Conn))
 
+	// To be registered later to trigger request dispatch
+	querier3Conn := NewUnregisteredQuerierWorkerConn(context.Background(), "querier-3")
+
 	t.Cleanup(func() {
 		// if the test has failed and the queue does not get cleared,
 		// we must send a shutdown signal for the remaining connected querier
 		// or else StopAndAwaitTerminated will never complete.
 		queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
+		queue.SubmitUnregisterQuerierWorkerConn(querier3Conn)
 		require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
 	})
+
+	assertControlFlowError := func(t *testing.T, err error) {
+		// Error received by the querier connections waiting for a dequeue should be one of:
+		// * ErrQuerierWorkerDisconnected: the waiting conn was crashed or otherwise deregistered but the whole querier is not yet deregistered
+		// * context.Canceled: context cancellation also propagates on waiting conn crash/deregister and races with ErrQuerierWorkerDisconnected
+		// * ErrQuerierShuttingDown: querier has initiated shutdown or all conns deregistered
+		// * ErrStopped: The scheduler has received a shutdown signal and cleared its queue
+		assert.True(t, errors.Is(err, ErrQuerierWorkerDisconnected) || errors.Is(err, context.Canceled) || errors.Is(err, ErrQuerierShuttingDown) || errors.Is(err, ErrStopped))
+	}
 
 	// Enqueue a request from a user which would NOT be assigned to querier-1.
 	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
@@ -384,66 +397,58 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	querier1DequeueReq := NewQuerierWorkerDequeueRequest(querier1Conn, FirstTenant())
 	go func() {
 		defer querier1SubmitDequeueReqWG.Done()
-		_, _, err = queue.AwaitRequestForQuerier(querier1DequeueReq)
-		// This receives one of two errors, it's a race between a context cancel and ErrQuerierWorkerDisconnected
-		require.Error(t, err)
+		_, _, err := queue.AwaitRequestForQuerier(querier1DequeueReq)
+		// This blocks until it receives one of the control flow errors;
+		// it will not receive a request as those in the queue are only sharded to querier-2.
+		assertControlFlowError(t, err)
 	}()
 
 	querier1CrashWG := sync.WaitGroup{}
 	querier1CrashWG.Add(1)
 	go func() {
 		defer querier1CrashWG.Done()
-		for queue.waitingDequeueRequestsToDispatch.Len() < 1 {
-			time.Sleep(time.Millisecond * 1)
-		}
-		// Simulate a crash of querier-1 (no graceful shutdown notification).
+		// Allow time for waiting querier dispatch request to be enqueued internally;
+		// we cannot check the queue length without triggering races but this is more than enough time.
+		time.Sleep(time.Millisecond * 100)
+
+		// Simulate a crash of querier-1's only worker connection (no graceful shutdown notification).
 		querier1CtxCancel(errors.New("crash"))
 		queue.SubmitUnregisterQuerierWorkerConn(querier1Conn) // normally done in the defers of the grpc loop
 	}()
 
-	// Enqueue another request from a user which would NOT be assigned to querier-1.
-	reqNotShardedToQuerier1 = &SchedulerRequest{
-		Ctx:                       context.Background(),
-		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
-		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
-	}
-	// The enqueuing of any new request triggers an attempt to dispatch requests
-	// to the items in waitingDequeueRequestsToDispatch list and their associated worker conns.
-	//
-	// The dequeue request at the front of that list will be associated with the disconnected worker conn,
-	// which now has a workerID set to unregisteredWorkerID = -1.
-	// Attempting to dequeue a request from the tree for an unregisteredWorkerID should fail, but not panic.
-	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
+	// Register another querier to trigger a re-shard and ensure we attempt to dispatch the requests.
+	// The queue will first try to dispatch a queue item to its internally-queued waiting dequeue requests.
+	// The first waiting dequeue request in that list is from the crashed and deregistered querier-1 -
+	// This test is to ensure that is handled gracefully.
+	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier3Conn))
+	querier3SubmitDequeueReqWG := sync.WaitGroup{}
+	querier3SubmitDequeueReqWG.Add(1)
+	go func() {
+		defer querier3SubmitDequeueReqWG.Done()
+		// This blocks until it receives one of the control flow errors;
+		// it will not receive a request as those in the queue are still only sharded to querier-2.
+		_, _, err := queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier3Conn, FirstTenant()))
+		assertControlFlowError(t, err)
+	}()
 
-	// The dequeue request should be dropped from the waiting list due to the error.
-	assert.Eventually(
+	// Deregister all remaining querier-workers to allow graceful shutdown without needing to clear queue.
+	queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
+	queue.SubmitUnregisterQuerierWorkerConn(querier3Conn)
+	// Shut down the queue so we can assert on its waiting dispatch requests without a data race
+	require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
+
+	// The dequeue request should have beeen dropped from the waiting list due to the error.
+	require.Eventually(
 		t,
 		func() bool { return queue.waitingDequeueRequestsToDispatch.Len() == 0 },
 		testTimeout,
 		1*time.Second,
 	)
 
-	// Allow Querier-2 to clear the queue after the disconnected querier-1 is forgotten and a reshard occurs.
-	querier2SubmitDequeueReqWG := sync.WaitGroup{}
-	querier2SubmitDequeueReqWG.Add(1)
-	go func() {
-		defer querier2SubmitDequeueReqWG.Done()
-
-		req, nextTenantIdx, err := queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier2Conn, FirstTenant()))
-		require.NoError(t, err)
-		require.NotNil(t, req)
-
-		req, _, err = queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier2Conn, nextTenantIdx))
-		require.NoError(t, err)
-		require.NotNil(t, req)
-	}()
-
-	querier2SubmitDequeueReqWG.Wait()
-	assert.True(t, queue.queueBroker.isEmpty())
-
 	// These should be long gone but ensure we have no hanging goroutines
 	querier1SubmitDequeueReqWG.Wait()
 	querier1CrashWG.Wait()
+	querier3SubmitDequeueReqWG.Wait()
 }
 
 func TestRequestQueue_RegisterAndUnregisterQuerierWorkerConnections(t *testing.T) {

--- a/pkg/scheduler/queue/queue_test.go
+++ b/pkg/scheduler/queue/queue_test.go
@@ -329,6 +329,13 @@ func queueConsume(
 	return lastTenantIdx, err
 }
 
+// TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker tests a scenario which previously caused a panic.
+// When a querier-worker submits a dequeue request while there are no queue items sharded to that querier,
+// The waiting dequeue request is held in an internal queue until a reshard or enqueue operation occurs.
+// A reshard or enqueue operation triggers an attempt to dispatch queue items to those waiting dequeue requests.
+//
+// If the querier-worker associated with the dequeue request has since crashed or otherwise been deregistered,
+// the queue should skip that dequeue request drop it from the internal queue so it will not be retried.
 func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.T) {
 	const forgetDelay = 2 * time.Second
 	const testTimeout = 10 * time.Second
@@ -376,12 +383,13 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 		assert.True(t, errors.Is(err, ErrQuerierWorkerDisconnected) || errors.Is(err, context.Canceled) || errors.Is(err, ErrQuerierShuttingDown) || errors.Is(err, ErrStopped))
 	}
 
-	// Enqueue requests from a user which would NOT be sharded to querier-1.
+	// Enqueue requests which will NOT be sharded to querier-1.
 	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
 	// when there are only one or two queriers in the sorted list of connected queriers.
 	//
-	// These will sit in the queue as user-1's sharded querier, querier-2 is not awaiting a request yet.
-	// Two different additional queue dimensions must exist in the queue tree to create the potential panic condition.
+	// These requests will sit in the queue -
+	// querier-2 is the only querier sharded to user-1, but querier-2 has not requested to dequeue yet.
+	// >1 queue dimensions must exist in the queue to reproduce a potential panic condition (dims % unregisteredWorkerID).
 	reqNotShardedToQuerier1 := &SchedulerRequest{
 		Ctx:                       context.Background(),
 		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},

--- a/pkg/scheduler/queue/queue_test.go
+++ b/pkg/scheduler/queue/queue_test.go
@@ -351,11 +351,11 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 	// Two queriers connect, allowing us to enqueue requests sharded to only one of the two.
 	querier1Ctx, querier1CtxCancel := context.WithCancelCause(context.Background())
 	querier1Conn := NewUnregisteredQuerierWorkerConn(querier1Ctx, "querier-1")
-	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier1Conn))
+	assert.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier1Conn))
 	querier2Conn := NewUnregisteredQuerierWorkerConn(context.Background(), "querier-2")
-	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier2Conn))
+	assert.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier2Conn))
 
-	// To be registered later to trigger request dispatch
+	// A third querier to be registered later to trigger request dispatch
 	querier3Conn := NewUnregisteredQuerierWorkerConn(context.Background(), "querier-3")
 
 	t.Cleanup(func() {
@@ -364,7 +364,7 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 		// or else StopAndAwaitTerminated will never complete.
 		queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
 		queue.SubmitUnregisterQuerierWorkerConn(querier3Conn)
-		require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
+		assert.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
 	})
 
 	assertControlFlowError := func(t *testing.T, err error) {
@@ -372,26 +372,31 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 		// * ErrQuerierWorkerDisconnected: the waiting conn was crashed or otherwise deregistered but the whole querier is not yet deregistered
 		// * context.Canceled: context cancellation also propagates on waiting conn crash/deregister and races with ErrQuerierWorkerDisconnected
 		// * ErrQuerierShuttingDown: querier has initiated shutdown or all conns deregistered
-		// * ErrStopped: The scheduler has received a shutdown signal and cleared its queue
+		// * ErrStopped: The scheduler has received a shutdown signal and cleared its queue or lost all querier connections.
 		assert.True(t, errors.Is(err, ErrQuerierWorkerDisconnected) || errors.Is(err, context.Canceled) || errors.Is(err, ErrQuerierShuttingDown) || errors.Is(err, ErrStopped))
 	}
 
-	// Enqueue a request from a user which would NOT be assigned to querier-1.
+	// Enqueue requests from a user which would NOT be sharded to querier-1.
 	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
-	// when there are only one or two queriers in the sorted list of connected queriers
+	// when there are only one or two queriers in the sorted list of connected queriers.
+	//
+	// These will sit in the queue as user-1's sharded querier, querier-2 is not awaiting a request yet.
+	// Two different additional queue dimensions must exist in the queue tree to create the potential panic condition.
 	reqNotShardedToQuerier1 := &SchedulerRequest{
 		Ctx:                       context.Background(),
 		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
-		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
+		AdditionalQueueDimensions: []string{"ingester"},
 	}
-	// This will sit in the queue as querier-2 is not await a request yet,
-	// but we need a request in the queue tree to reach the code path
-	// in QuerierWorkerQueuePriorityAlgo.setup where the panic condition is created
-	// by setting the currentNodeOrderIndex to -1.
-	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
+	assert.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
+	reqNotShardedToQuerier1 = &SchedulerRequest{
+		Ctx:                       context.Background(),
+		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
+		AdditionalQueueDimensions: []string{"store-gateway"},
+	}
+	assert.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotShardedToQuerier1, 1, nil))
 
 	// Querier-1 submits a request to dequeue;
-	// it will not receive anything as the only request in the queue is not sharded to querier-1.
+	// it will not receive anything as the only requests in the queue are not sharded to querier-1.
 	querier1SubmitDequeueReqWG := sync.WaitGroup{}
 	querier1SubmitDequeueReqWG.Add(1)
 	querier1DequeueReq := NewQuerierWorkerDequeueRequest(querier1Conn, FirstTenant())
@@ -403,52 +408,39 @@ func TestDispatchToWaitingDequeueRequestForUnregisteredQuerierWorker(t *testing.
 		assertControlFlowError(t, err)
 	}()
 
-	querier1CrashWG := sync.WaitGroup{}
-	querier1CrashWG.Add(1)
-	go func() {
-		defer querier1CrashWG.Done()
-		// Allow time for waiting querier dispatch request to be enqueued internally;
-		// we cannot check the queue length without triggering races but this is more than enough time.
-		time.Sleep(time.Millisecond * 100)
+	// Wait for the waiting dequeue request to be enqueued internally.
+	for queue.waitingDequeueRequestsToDispatchCount.Load() < 1 {
+		time.Sleep(1 * time.Millisecond)
+	}
 
-		// Simulate a crash of querier-1's only worker connection (no graceful shutdown notification).
-		querier1CtxCancel(errors.New("crash"))
-		queue.SubmitUnregisterQuerierWorkerConn(querier1Conn) // normally done in the defers of the grpc loop
-	}()
+	// Simulate a crash of querier-1's only connection, with no graceful shutdown notification from the querier.
+	// The entire querier is not removed until forgetDelay passes, but the individual connection is marked as unregistered.
+	// A waiting dequeue request must have its connection ID set to -1 to complete the potential panic condition.
+	querier1CtxCancel(errors.New("crash"))
+	queue.SubmitUnregisterQuerierWorkerConn(querier1Conn) // normally done in the defers of the grpc loop
+
+	// Wait for the queue to process the de-registration of the querier-1 connection.
+	for queue.connectedQuerierWorkers.Load() > 1 {
+		time.Sleep(1 * time.Millisecond)
+	}
 
 	// Register another querier to trigger a re-shard and ensure we attempt to dispatch the requests.
 	// The queue will first try to dispatch a queue item to its internally-queued waiting dequeue requests.
 	// The first waiting dequeue request in that list is from the crashed and deregistered querier-1 -
 	// This test is to ensure that is handled gracefully.
-	require.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier3Conn))
-	querier3SubmitDequeueReqWG := sync.WaitGroup{}
-	querier3SubmitDequeueReqWG.Add(1)
-	go func() {
-		defer querier3SubmitDequeueReqWG.Done()
-		// This blocks until it receives one of the control flow errors;
-		// it will not receive a request as those in the queue are still only sharded to querier-2.
-		_, _, err := queue.AwaitRequestForQuerier(NewQuerierWorkerDequeueRequest(querier3Conn, FirstTenant()))
-		assertControlFlowError(t, err)
-	}()
+	assert.NoError(t, queue.AwaitRegisterQuerierWorkerConn(querier3Conn))
 
-	// Deregister all remaining querier-workers to allow graceful shutdown without needing to clear queue.
-	queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
-	queue.SubmitUnregisterQuerierWorkerConn(querier3Conn)
-	// Shut down the queue so we can assert on its waiting dispatch requests without a data race
-	require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
-
-	// The dequeue request should have beeen dropped from the waiting list due to the error.
-	require.Eventually(
+	// The dequeue request should have been dropped from the waiting list
+	// when the queue received an error from the queue broker's dequeueRequestForQuerier.
+	assert.Eventually(
 		t,
-		func() bool { return queue.waitingDequeueRequestsToDispatch.Len() == 0 },
+		func() bool { return queue.waitingDequeueRequestsToDispatchCount.Load() == 0 },
 		testTimeout,
 		1*time.Second,
 	)
 
-	// These should be long gone but ensure we have no hanging goroutines
+	// This should be long gone but ensure we have no hanging goroutines
 	querier1SubmitDequeueReqWG.Wait()
-	querier1CrashWG.Wait()
-	querier3SubmitDequeueReqWG.Wait()
 }
 
 func TestRequestQueue_RegisterAndUnregisterQuerierWorkerConnections(t *testing.T) {

--- a/pkg/scheduler/scheduler.go
+++ b/pkg/scheduler/scheduler.go
@@ -442,6 +442,11 @@ func (s *Scheduler) QuerierLoop(querier schedulerpb.SchedulerForQuerier_QuerierL
 		dequeueReq := queue.NewQuerierWorkerDequeueRequest(querierWorkerConn, lastTenantIdx)
 		queryReq, idx, err := s.requestQueue.AwaitRequestForQuerier(dequeueReq)
 		if err != nil {
+			// The error returned can either be ErrQuerierShuttingDown or ErrQuerierWorkerDisconnected.
+			// ErrQuerierShuttingDown is a control-flow message between services to exit this loop.
+			// (see note in transformRequestQueueError).
+			// ErrQuerierWorkerDisconnected is caused by connection/goroutine crash;
+			// we expect this loop is no longer alive to receive the error anyway.
 			return s.transformRequestQueueError(err)
 		}
 

